File: examples/concepts/cpp/methods/Predicates.cpp
// -*- tab-width:2 ; indent-tabs-mode:nil -*-
//:: cases Predicates
//:: tools silicon
//:: verdict Pass

int amount = 0;

//@ resource pre_state() = Perm(amount, write);
//@ resource state(int value) = Perm(amount, write) ** amount == value;

//@ requires pre_state();
//@ ensures state(x);
void set(int x) {
	// pre_state();
	//@ unfold pre_state();
	// Perm(amount, write);
	amount = x;
	// Perm(amount, write) ** amount == x;
	//@ fold state(x);
	// state(x);
}

//@ requires pre_state();
//@ ensures state(50);
void manageAccount() {
	set(50);
}

File: examples/runtime/predicates.java
public class Main {

    //@ requires c != null ** c.state(0);
    //@ ensures c.state(2);
    void foo(Counter c) {
        c.increment(2);
    }

    public void execute() {
        Counter c = new Counter();
        //@ fold c.state(0);
        Counter c2 = new Counter();
        //@ fold c2.state(0);

        foo(c);
        foo(c2);
    }

    public void main() {
        Main m = new Main();
        m.execute();
    }
}


class Counter {
    int count;

    //@ resource state(int val) = Perm(count, write) ** count == val;

    //@ requires state(count);
    //@ ensures state(count);
    void increment(int n) {
        int z = count;
        //@ unfold state(count);
        count += n;
        //@ fold state(count);
    };
}